open-build-service-api/tests/integration.rs

```diff
@@ -880,6 +880,153 @@ async fn test_build_status() {
     assert!(status.dirty);
 }
 
+#[tokio::test]
+async fn test_build_rebuild() {
+    let mock = start_mock().await;
+
+    mock.add_project(TEST_PROJECT.to_owned());
+    mock.add_new_package(
+        TEST_PROJECT,
+        TEST_PACKAGE_1.to_owned(),
+        MockPackageOptions::default(),
+    );
+    mock.add_new_package(
+        TEST_PROJECT,
+        TEST_PACKAGE_2.to_owned(),
+        MockPackageOptions::default(),
+    );
+
+    mock.add_or_update_repository(
+        TEST_PROJECT,
+        TEST_REPO.to_owned(),
+        TEST_ARCH_1.to_owned(),
+        MockRepositoryCode::Building,
+    );
+
+    mock.set_package_build_status(
+        TEST_PROJECT,
+        TEST_REPO,
+        TEST_ARCH_1,
+        TEST_PACKAGE_1.to_owned(),
+        MockBuildStatus::new(MockPackageCode::Blocked),
+    );
+    mock.set_package_build_status(
+        TEST_PROJECT,
+        TEST_REPO,
+        TEST_ARCH_1,
+        TEST_PACKAGE_2.to_owned(),
+        MockBuildStatus::new(MockPackageCode::Blocked),
+    );
+    mock.set_package_build_status_for_rebuilds(
+        TEST_PROJECT,
+        MockBuildStatus::new(MockPackageCode::Building),
+    );
+
+    let obs = create_authenticated_client(mock.clone());
+
+    let status = obs
+        .project(TEST_PROJECT.to_owned())
+        .package(TEST_PACKAGE_1.to_owned())
+        .status(TEST_REPO, TEST_ARCH_1)
+        .await
+        .unwrap();
+    assert_eq!(status.code, PackageCode::Blocked);
+
+    obs.project(TEST_PROJECT.to_owned())
+        .package(TEST_PACKAGE_1.to_owned())
+        .rebuild()
+        .await
+        .unwrap();
+
+    let status = obs
+        .project(TEST_PROJECT.to_owned())
+        .package(TEST_PACKAGE_1.to_owned())
+        .status(TEST_REPO, TEST_ARCH_1)
+        .await
+        .unwrap();
+    assert_eq!(status.code, PackageCode::Building);
```

Comment:
I'm admittedly not sure what you mean here :sweat_smile: I could certainly try and emulate it more precisely in the mocking layer, but I wasn't that familiar with how it works and didn't want to hardcode in a bunch of mistaken assumption.

IMO though, being able to set custom build statuses here is just a bit more convenient, I'm not entirely sure under what circumstances actually emulating the full flow here would be particularly beneficial, other than just making it a bit annoying to configure? Since then, the user of the mock APIs would have to have some way to configure where & when to go from "scheduled" too?

Reply:
Don't think it's critical; I'm a bit worried about other crates (say the runner) using this for mocking but the state machine not being the same (for as far as there is a real state machine). But i guess we can deal with that later